Bi-directional dshot on iomcu F103 8MHz

burgeruser opened the pull request:
This PR adds support for bdshot telemetry on iomcus using the F103 with 8Mhz. There is one limitation - only channels 1-4,7-8 support telemetry. It is not possible to support 5-6 as these share a DMA channel with the fast UART interconnect.

Support is built into the iomcu dshot variant so to use you simply need to set BRD_IO_DSHOT and supply the appropriate SERVO_BLH_BDMASK. ESC telemetry indexes are modified to reflect the fact that ESC's 1-4 and 7-8 can now provide telemetry.

EDT is working, so you can get temperature, current and voltage information from the ESCs using bdshot as well. To use set SERVO_DSHOT_ESC in the normal way.

The two binary files named "iofirmware_f103_8MHz_dshot_lowpolh.bin" and iofirmware_f103_8MHz_dshot_highpolh.bin"  were compiled using **Copter-4.5.2**.

andyp1per reviewed (changes requested):
Very nice!

I think its better to put a hwdef.inc in iomcu-f103-dshot and then reference that from both places in order to avoid code duplication

burgeruser commented:
> Very nice!
> 
> I think its better to put a hwdef.inc in iomcu-f103-dshot and then reference that from both places in order to avoid code duplication

I've just tested on **Podman H7** flight controller, and everything is currently working as expected.

andyp1per commented:
Great!

Still need the inc change

burgeruser commented:
> Great!
> 
> Still need the inc change
this 

I think its better to put a hwdef.inc in iomcu-f103-dshot and then reference that from both places in order to avoid code duplication.

Before:
![before](https://github.com/ArduPilot/ardupilot/assets/93718551/7f1705b2-c6b0-4e0f-80a1-233b5849fcfb)


After I have changed:
![after](https://github.com/ArduPilot/ardupilot/assets/93718551/4a8901b7-03b5-45ef-9236-4e66c0d5bd9c)

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/iomcu-f103-dshot/hwdef.inc:
```diff
@@ -0,0 +1,93 @@
+# hw definition file for processing by chibios_pins.py
+
+include ../iomcu/hwdef.inc
```
This needs to almost exactly match the old hwdef.dat. You have added a whole bunch of extra lines - please remove them

burgeruser replied:
removed them

burgeruser replied:
> This needs to almost exactly match the old hwdef.dat. You have added a whole bunch of extra lines - please remove them
remove them

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/iomcu-f103-8Mhz-dshot/hwdef.inc:
```diff
@@ -0,0 +1,98 @@
+
+include ../iomcu/hwdef.inc
+
+# defines to allow the loop timing to be observed with a Saleae
```
This should include iomcu-f103-dshot/hwdef.inc and almost all the lines should be deleted

andyp1per replied:
Nothing actually includes this so I don't think there is a need for another hwdef.inc - just use hwdef.dat

burgeruser replied:
You are right - just use hwdef.dat in iomcu-f103-8Mhz-dshot/hwdef.dat.

burgeruser commented:
I recently made modifications to the file "iomcu-f103-dshot/hwdef.inc". 
When editing, I aimed to follow the previous coding style of Ardupilot and maintained the order of define consistently, ensuring clarity for readers. 
As a result, redundant .inc suffix files should no longer be necessary.
I have deleted all the lines should be deleted in file "iomcu-f103-dshot/hwdef.inc",and modified "iomcu-f103-dshot/hwdef.dat", "iomcu-f103-8Mhz-dshot/hwdef.dat" files,  just use hwdef.dat in iomcu-f103-8Mhz-dshot/hwdef.dat.

burgeruser commented:
I recently made modifications to the file "iomcu-f103-dshot/hwdef.inc".
When editing, I aimed to follow the previous coding style of Ardupilot and maintained the order of define consistently, ensuring clarity for readers.
As a result, redundant .inc suffix files should no longer be necessary.
I have deleted all the lines should be deleted in file "iomcu-f103-dshot/hwdef.inc",and modified "iomcu-f103-dshot/hwdef.dat", "iomcu-f103-8Mhz-dshot/hwdef.dat" files, just use hwdef.dat in iomcu-f103-8Mhz-dshot/hwdef.dat.
![after](https://github.com/ArduPilot/ardupilot/assets/93718551/7266e63b-4dc1-439e-80e1-04780c7bf265)

andyp1per commented:
I've fixed it for you - please check it

burgeruser commented:
> I've fixed it for you - please check it

I understand that code style encourages **minimizing redundant code** by leveraging encapsulation. Key to this principle is the use of functions, methods, classes, or modules, which allow for **reusability**. `Include` or `import` statements are commonly employed to incorporate external code or modules, facilitating the reuse of existing functionalities, thereby reducing the likelihood of errors and enhancing development efficiency.

burgeruser commented:
I am currently attempting to test with an octocopter.

burgeruser commented:
I understand that code style encourages minimizing redundant code by leveraging encapsulation. Key to this principle is the use of functions, methods, classes, or modules, which allow for reusability. Include or import statements are commonly employed to incorporate external code or modules, facilitating the reuse of existing functionalities, thereby reducing the likelihood of errors and enhancing development efficiency.
![after](https://github.com/ArduPilot/ardupilot/assets/93718551/31e6c863-8701-4215-9c35-bf3b1029fb65)

burgeruser commented:
The two binary files named "iofirmware_f103_8MHz_dshot_lowpolh.bin" and iofirmware_f103_8MHz_dshot_highpolh.bin" were compiled using Copter-4.5.2.
After a careful compilation on my computer, I successfully tested with the **Podman H7 flight controller** using the iomcu output for the 8-channel DShot signal on an octocopter. The results were flawless, allowing for proper motor speed feedback.
![test](https://github.com/ArduPilot/ardupilot/assets/93718551/73fa5c54-b6e2-40b7-aa62-5f20e0e501d1)

andyp1per commented:
The build script needed updating and I also needed to make sure that the binaries were compiled with the right compiler - so I have fixed this now. Please check it.

andyp1per reviewed (changes requested):
Commits now broken

andyp1per commented:
You now need to squash and merge your commit into the other one

burgeruser commented:
I recently standardized the folder names to **8MHz**, ensuring no compilation errors. Earlier issues arose due to inconsistent naming in the path folders, specifically "iomcu-f103-8Mhz-dshot". I addressed the Python file validation issue related to this, and now, I've resolved the root problem by renaming the folders to "iomcu-f103-8MHz-dshot".

andyp1per reviewed (changes requested):
You should change everything to use MHz, both to be consistent with the other directory and because that is the correct capitalisation

burgeruser commented:
I will squash and merge your commit into the other one

burgeruser commented:
> You now need to squash and merge your commit into the other one

 This time, my modifications were based on the latest **master branch**. I then compressed and merged all my commits into a single new commit.

andyp1per commented:
You need to learn how to use git. I have redone the commits for you so that they have the right messages and are split appropriately. Please don't change anything now.

burgeruser commented:
> You need to learn how to use git. I have redone the commits for you so that they have the right messages and are split appropriately. Please don't change anything now. I have also rebuilt the firmware so that it is using the right branch and compiler.

Thank you very much, this is my first successful merge with Ardupilot's code, it's been a fantastic experience.
